Set memfd for memorybacking

alicefr opened the pull request:
**What this PR does / why we need it**:
This PR sets as default memfd as source type for memorybacking when hugetables are used.

Fixes #3781

I still have various questions and I'm not sure how to handle it:

1. Do we want to use `memfd` as default or only in certain cases?
2. NUMA is required in order to use `memfd`. If it is not set, libvirt is simply using `/dev/hugetables`. How do we want to generate the NUMA configuration? Which components should generate this?

About the 2nd point, right now I hacked the NUMA generation based on the number of VCPUs and the memory. I don't think that this is correct.

For the NUMA definition, I cherry-picked one of the commits (https://github.com/kubevirt/kubevirt/pull/3493/commits/09fd983c97bb9c1c312fa90fb90e4ca35ecf95d7)  from https://github.com/kubevirt/kubevirt/pull/3493

/cc @vladikr 

**Special notes for your reviewer**:
I tested this PR using this yaml `vmi-hugepages.yaml`:
```yaml
---
apiVersion: kubevirt.io/v1alpha3
kind: VirtualMachineInstance
metadata:
  labels:
    special: vmi-hugepages
  name: vmi-hugepages
spec:
  domain:
    cpu:
      dedicatedCpuPlacement: true
    memory:
      hugepages:
        pageSize: 2Mi
    devices:
      disks:
      - disk:
          bus: virtio
        name: containerdisk
    machine:
      type: ""
    resources:
      requests:
        memory: 64Mi
      limits:
        cpu: "2"
        memory: 64Mi
  terminationGracePeriodSeconds: 0
  volumes:
  - containerDisk:
      image: registry:5000/kubevirt/cirros-container-disk-demo:devel
    name: containerdisk

```
Steps to reproduce my setup:
```bash
$ make
$ make cluster-up && make cluster-sync
$ cluster-up/kubectl.sh label --overwrite no node01 cpumanager=true
$ cluster-up/kubectl.sh apply -f vmi-hugepages.yaml
$ cluster-up/kubectl.sh logs -c compute $(cluster-up/kubectl.sh get pods | grep vmi | awk '{print $1}') | grep qemu 
[...]
\\-object memory-backend-memfd,id=ram-node0,hugetlb=yes,hugetlbsize=2097152,size=67108864 \\-numa node,nodeid=0,cpus=0-1,memdev=ram-node0

```
**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Added memfd as default memory backend for hugepages. This introduces the new annotation kubevirt.io/memfd to disable memfd as default and fallback to the previous behavior.
```

alicefr commented:
/cc @crobinso

vladikr commented:
Thank you for this PR @alicefr !


> I still have various questions and I'm not sure how to handle it:
> 
> 1. Do we want to use `memfd` as default or only in certain cases?

I think we should just use it as a default. It removes the need for additional selinux privileges we have to carry for hugepages .

> 2. NUMA is required in order to use `memfd`. If it is not set, libvirt is simply using `/dev/hugetables`. How do we want to generate the NUMA configuration? Which components should generate this?
> 
> About the 2nd point, right now I hacked the NUMA generation based on the number of VCPUs and the memory. I don't think that this is correct.

Yes. I also did the same in @3493 - I think it should be "good enough" at this time. We don't have support for NUMA on the guest side - so creating a single numa node with all the available cpus and memory seems correct to me. We will need to enhance this going forward of course.

> For the NUMA definition, I cherry-picked one of the commits ([09fd983](https://github.com/kubevirt/kubevirt/commit/09fd983c97bb9c1c312fa90fb90e4ca35ecf95d7)) from #3493
:+1: 

Thanks!

vladikr commented:
/cc @rmohr @davidvossel

alicefr commented:
@vladikr thanks!

> > About the 2nd point, right now I hacked the NUMA generation based on the number of VCPUs and the memory. I don't think that this is correct.
> 
> Yes. I also did the same in #3493 - I think it should be "good enough" at this time. We don't have support for NUMA on the guest side - so creating a single numa node with all the available cpus and memory seems correct to me. We will need to enhance this going forward of course.

Ok, if it's acceptable to generate the NUMA configuration in this way, there's a small issue to be fix. Right know, I'm reading the VCPU and this is set with:
```yaml
    resources:
      limits:
        cpu: "2"
```
If this doesn't become a "mandatory" field, then if it is not set, I need to get the total number of cpus. I'll check from where :)

vladikr commented:
> @vladikr thanks!
> 
> > > About the 2nd point, right now I hacked the NUMA generation based on the number of VCPUs and the memory. I don't think that this is correct.
> > 
> > 
> > Yes. I also did the same in #3493 - I think it should be "good enough" at this time. We don't have support for NUMA on the guest side - so creating a single numa node with all the available cpus and memory seems correct to me. We will need to enhance this going forward of course.
> 
> Ok, if it's acceptable to generate the NUMA configuration in this way, there's a small issue to be fixed. Right know, I'm reading the VCPU and this is set with:
> 
> ```yaml
>     resources:
>       limits:
>         cpu: "2"
> ```
> 
> If this doesn't become a "mandatory" field, then if it is not set, I need to get the total number of cpus. I'll check from where :)

@alicefr yes, vcpus may have different sources. Please take a look at what I did in https://github.com/kubevirt/kubevirt/pull/3493/files#diff-caee6d384498fde52fde15891ac547aaR660

alicefr commented:
@vladikr do you mind if we separate the NUMA set from this and virtiofs PR #3493 and we set it whenever hugetables are used?

vladikr commented:
/retest

vladikr commented:
/retest

alicefr commented:
I probably need to fix the memory assignment. The tests are not passing because a mismatch of the expected value and the real one

alicefr commented:
We need to calculate differently the expected memory once we switch to memfd: https://github.com/kubevirt/kubevirt/blob/master/tests/vmi_configuration_test.go#L950-L955. Just to figure out the right way

alicefr commented:
I'm not sure if the memory calculation [here](https://github.com/kubevirt/kubevirt/blob/master/tests/vmi_configuration_test.go#L950-L956) is correct. It doesn't take into account if there are any reserved hugepages. I'm trying to understand why the tests are failing. So, I tried to deploy a small example that request 32Mi and the pagsize is 2Mi:
```yaml
---
apiVersion: kubevirt.io/v1alpha3
kind: VirtualMachineInstance
metadata:
  labels:
    special: vmi-hugepages
  name: vmi-hugepages
spec:
  domain:
    memory:
      hugepages:
        pageSize: 2Mi
    devices:
      disks:
      - disk:
          bus: virtio
        name: containerdisk
    machine:
      type: ""
    resources:
      requests:
        memory: 32Mi
  terminationGracePeriodSeconds: 0
  volumes:
  - containerDisk:
      image: registry:5000/kubevirt/cirros-container-disk-demo:devel
```
So, I should get 16 free hugepages less. Instead, it's what I got:
```bash
cat /proc/meminfo | grep -i hugepage
AnonHugePages:    716800 kB
ShmemHugePages:        0 kB
HugePages_Total:      64
HugePages_Free:       54
HugePages_Rsvd:        7
HugePages_Surp:        0
Hugepagesize:       2048 kB

```

rmohr commented:
@alicefr where do you get these values from? From inside the guest or from inside the container?

alicefr commented:
@rmohr from outside the container. Good point! Let me try inside the container.

alicefr commented:
@rmohr from inside the container:
```bash
$  cluster-up/kubectl.sh exec -ti -c compute virt-launcher-vmi-hugepages-fzmht -- bash
[root@vmi-hugepages /]# cat /proc/meminfo | grep -i hugepage
AnonHugePages:    733184 kB
ShmemHugePages:        0 kB
HugePages_Total:      64
HugePages_Free:       54
HugePages_Rsvd:        7
HugePages_Surp:        0
Hugepagesize:       2048 kB

```

alicefr commented:
With 64Mi memory (from inside the container):
```bash
$ cat /proc/meminfo | grep -i hugepage
AnonHugePages:    743424 kB
ShmemHugePages:        0 kB
HugePages_Total:      64
HugePages_Free:       31
HugePages_Rsvd:        1
HugePages_Surp:        0
Hugepagesize:       2048 kB

```

rmohr commented:
@alicefr I think that you have to check the cgroups of the pods. Afair pods see all hugepages but can only consume the requested amount. Could you also check the pod spec? We request all hugepages through k8s, so I would expect that the cgroups match what we request on the pod.

alicefr commented:
@rmohr yes, I can. However, I'm just reproducing what the test is doing [here](https://github.com/kubevirt/kubevirt/blob/master/tests/vmi_configuration_test.go#L950-L956), it is reading the values from sysfs here: https://github.com/kubevirt/kubevirt/blob/5f6375ca86815f50274cdd60f9153bd13847d70a/tests/vmi_configuration_test.go#L925 
 
In my example, I used `/proc/meminfo`, but the values are the same that in sysfs.
In `compute` container:
```bash      
[root@vmi-hugepages /]$ cat /sys/kernel/mm/hugepages/hugepages-2048kB/nr_hugepages
64
[root@vmi-hugepages /]$ cat /sys/kernel/mm/hugepages/hugepages-2048kB/free_hugepages 
54  
[root@vmi-hugepages /]$ cat /sys/kernel/mm/hugepages/hugepages-2048kB/resv_hugepages 
7

```

alicefr commented:
Sorry, I've also to correct myself:). I mean that the calculation `nr_hugepages - free_hugepages`  might be not corrected anymore because with memfd as memory backend also `resv_hugepages ` are used

alicefr commented:
/retest

vladikr commented:
/test all

alicefr commented on pkg/virt-launcher/virtwrap/api/converter.go:
```diff
@@ -816,7 +816,21 @@ func Convert_v1_VirtualMachine_To_api_Domain(vmi *v1.VirtualMachineInstance, dom
+				{
+					ID:     "0",
+					CPUs:   fmt.Sprintf("0-%d", domain.Spec.VCPU.CPUs-1),
+					Memory: fmt.Sprintf("%d", getVirtualMemory(vmi).ScaledValue(resource.Mega)),
```
This is still wrong. I need to assign correctly the unit

alicefr replied:
I tried to modify in this way:
```diff
diff --git a/pkg/virt-launcher/virtwrap/api/converter.go b/pkg/virt-launcher/virtwrap/api/converter.go
index 270b27b41..6f3328080 100644
--- a/pkg/virt-launcher/virtwrap/api/converter.go
+++ b/pkg/virt-launcher/virtwrap/api/converter.go
@@ -825,8 +825,8 @@ func Convert_v1_VirtualMachine_To_api_Domain(vmi *v1.VirtualMachineInstance, dom
                                {
                                        ID:     "0",
                                        CPUs:   fmt.Sprintf("0-%d", domain.Spec.VCPU.CPUs-1),
-                                       Memory: fmt.Sprintf("%d", getVirtualMemory(vmi).ScaledValue(resource.Mega)),
-                                       Unit:   "MiB",
+                                       Memory: fmt.Sprintf("%d", getVirtualMemory(vmi).Value()/int64(1024)),
+                                       Unit:   "KiB",
                                },
                        },
                }
```
but the GiB test is still failing with:
```
------------------------------
• Failure [13.177 seconds]
Configurations
/root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:56
  VirtualMachineInstance definition
  /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:172
    [rfe_id:140][crit:medium][vendor:cnv-qe@redhat.com][level:component]with hugepages
    /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:912
      should consume hugepages 
      /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table.go:92
        [test_id:1672]hugepages-1Gi [It]
        /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table_entry.go:43

        Expected
            <int64>: 65011712
        to equal
            <int64>: 67108864

```

vladikr replied:
It's getting closer. Let me pull your PR and run it locally.

alicefr replied:
My suspicion is that the test is failing because the system is not totally clean up :/.
I just tried to run separately the tests:
Test 2Mi:
```bash
$ FUNC_TEST_ARGS='-ginkgo.focus=hugepages-2Mi' make functest
hack/dockerized "hack/build-func-tests.sh"
go version go1.12.8 linux/amd64

go version go1.12.8 linux/amd64
hack/functests.sh
Running Suite: Tests Suite
==========================
Random Seed: 1598515325
Will run 1 of 577 specs

SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
• [SLOW TEST:13.076 seconds]
Configurations
/root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:56
  VirtualMachineInstance definition
  /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:172
    [rfe_id:140][crit:medium][vendor:cnv-qe@redhat.com][level:component]with hugepages
    /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:912
      should consume hugepages 
      /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table.go:92
        [test_id:1671]hugepages-2Mi
        /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table_entry.go:43
------------------------------
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
Waiting for namespace kubevirt-test-default to be removed, this can take a while ...
Waiting for namespace kubevirt-test-alternative to be removed, this can take a while ...
Waiting for namespace kubevirt-test-operator to be removed, this can take a while ...

Ran 1 of 577 Specs in 25.413 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 576 Skipped
PASS

```
Test 1Gi:
```bash
$ FUNC_TEST_ARGS='-ginkgo.focus=hugepages-1Gi' make functest
hack/dockerized "hack/build-func-tests.sh"
go version go1.12.8 linux/amd64

go version go1.12.8 linux/amd64
hack/functests.sh
Running Suite: Tests Suite
==========================
Random Seed: 1598515143
Will run 2 of 577 specs

SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
S [SKIPPING] [4.607 seconds]
Configurations
/root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:56
  VirtualMachineInstance definition
  /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:172
    [rfe_id:140][crit:medium][vendor:cnv-qe@redhat.com][level:component]with hugepages
    /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:912
      should consume hugepages 
      /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table.go:92
        [test_id:1672]hugepages-1Gi [It]
        /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table_entry.go:43

        No node with hugepages hugepages-1Gi capacity

        /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:977
------------------------------
• [SLOW TEST:12.022 seconds]
Configurations
/root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:56
  VirtualMachineInstance definition
  /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:172
    [rfe_id:140][crit:medium][vendor:cnv-qe@redhat.com][level:component]with hugepages
    /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:912
      should consume hugepages 
      /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table.go:92
        [test_id:1672]hugepages-1Gi
        /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table_entry.go:43
------------------------------
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
Waiting for namespace kubevirt-test-default to be removed, this can take a while ...
Waiting for namespace kubevirt-test-alternative to be removed, this can take a while ...
Waiting for namespace kubevirt-test-operator to be removed, this can take a while ...

Ran 1 of 577 Specs in 48.889 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 576 Skipped
PASS

```

All together:
```bash
FUNC_TEST_ARGS='-ginkgo.focus=hugepages' make functest
hack/dockerized "hack/build-func-tests.sh"
go version go1.12.8 linux/amd64

go version go1.12.8 linux/amd64
hack/functests.sh
Running Suite: Tests Suite
==========================
Random Seed: 1598515433
Will run 4 of 577 specs

SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
• [SLOW TEST:11.003 seconds]
Configurations
/root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:56
  VirtualMachineInstance definition
  /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:172
    [rfe_id:140][crit:medium][vendor:cnv-qe@redhat.com][level:component]with hugepages
    /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:912
      should consume hugepages 
      /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table.go:92
        [test_id:1671]hugepages-2Mi
        /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table_entry.go:43
------------------------------
S [SKIPPING] [2.603 seconds]
Configurations
/root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:56
  VirtualMachineInstance definition
  /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:172
    [rfe_id:140][crit:medium][vendor:cnv-qe@redhat.com][level:component]with hugepages
    /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:912
      should consume hugepages 
      /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table.go:92
        [test_id:1672]hugepages-1Gi [It]
        /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table_entry.go:43

        No node with hugepages hugepages-1Gi capacity

        /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:977
------------------------------
• Failure [12.079 seconds]
Configurations
/root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:56
  VirtualMachineInstance definition
  /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:172
    [rfe_id:140][crit:medium][vendor:cnv-qe@redhat.com][level:component]with hugepages
    /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:912
      should consume hugepages 
      /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table.go:92
        [test_id:1672]hugepages-1Gi [It]
        /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table_entry.go:43

        Expected
            <int64>: 65011712
        to equal
            <int64>: 67108864

        /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:965
------------------------------
STEP: Starting a VM
{"component":"tests","kind":"VirtualMachineInstance","level":"info","msg":"Event(v1.ObjectReference{Kind:\"VirtualMachineInstance\", Namespace:\"kubevirt-test-default\", Name:\"testvmiv4c8g97959b4fx9nvktcxvxjzpzj89gnnp6mxzd5fclx5gwp\", UID:\"bf7de901-b744-492a-bda0-0d058ddc6466\", APIVersion:\"kubevirt.io/v1alpha3\", ResourceVersion:\"20095\", FieldPath:\"\"}): type: 'Normal' reason: 'SuccessfulCreate' Created virtual machine pod virt-launcher-testvmiv4c8g97959b4fx9nvktcxvxjzpzj89gnnp6mxj7q5f","name":"testvmiv4c8g97959b4fx9nvktcxvxjzpzj89gnnp6mxzd5fclx5gwp","namespace":"kubevirt-test-default","pos":"utils.go:327","timestamp":"2020-08-27T08:04:13.945889Z","uid":"bf7de901-b744-492a-bda0-0d058ddc6466"}
{"component":"tests","kind":"VirtualMachineInstance","level":"info","msg":"Event(v1.ObjectReference{Kind:\"VirtualMachineInstance\", Namespace:\"kubevirt-test-default\", Name:\"testvmiv4c8g97959b4fx9nvktcxvxjzpzj89gnnp6mxzd5fclx5gwp\", UID:\"bf7de901-b744-492a-bda0-0d058ddc6466\", APIVersion:\"kubevirt.io/v1alpha3\", ResourceVersion:\"20155\", FieldPath:\"\"}): type: 'Normal' reason: 'Created' VirtualMachineInstance defined.","name":"testvmiv4c8g97959b4fx9nvktcxvxjzpzj89gnnp6mxzd5fclx5gwp","namespace":"kubevirt-test-default","pos":"utils.go:327","timestamp":"2020-08-27T08:04:20.091882Z","uid":"bf7de901-b744-492a-bda0-0d058ddc6466"}
{"component":"tests","kind":"VirtualMachineInstance","level":"info","msg":"Event(v1.ObjectReference{Kind:\"VirtualMachineInstance\", Namespace:\"kubevirt-test-default\", Name:\"testvmiv4c8g97959b4fx9nvktcxvxjzpzj89gnnp6mxzd5fclx5gwp\", UID:\"bf7de901-b744-492a-bda0-0d058ddc6466\", APIVersion:\"kubevirt.io/v1alpha3\", ResourceVersion:\"20160\", FieldPath:\"\"}): type: 'Normal' reason: 'Started' VirtualMachineInstance started.","name":"testvmiv4c8g97959b4fx9nvktcxvxjzpzj89gnnp6mxzd5fclx5gwp","namespace":"kubevirt-test-default","pos":"utils.go:327","timestamp":"2020-08-27T08:04:20.134800Z","uid":"bf7de901-b744-492a-bda0-0d058ddc6466"}
STEP: Checking that the VM memory equals to a number of consumed hugepages
failed to dump entity named [sriovnetworknodestates]: the server could not find the requested resource
failed to dump entity named [sriovnetworknodepolicies]: the server could not find the requested resource
failed to dump entity named [sriovnetworks]: the server could not find the requested resource
failed to dump entity named [sriovoperatorconfigs]: the server could not find the requested resource
failed to dump entity named [network-attachment-definitions]: the server could not find the requested resource
• [SLOW TEST:5.635 seconds]
Configurations
/root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:56
  VirtualMachineInstance definition
  /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:172
    [rfe_id:140][crit:medium][vendor:cnv-qe@redhat.com][level:component]with hugepages
    /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:912
      with unsupported page size
      /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:1016
        [test_id:1673]should failed to schedule the pod
        /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:1017
------------------------------
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
Waiting for namespace kubevirt-test-default to be removed, this can take a while ...
Waiting for namespace kubevirt-test-alternative to be removed, this can take a while ...
Waiting for namespace kubevirt-test-operator to be removed, this can take a while ...


Summarizing 1 Failure:

[Fail] Configurations VirtualMachineInstance definition [rfe_id:140][crit:medium][vendor:cnv-qe@redhat.com][level:component]with hugepages should consume hugepages  [It] [test_id:1672]hugepages-1Gi 
/root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:965

Ran 3 of 577 Specs in 56.218 seconds
FAIL! -- 2 Passed | 1 Failed | 0 Pending | 574 Skipped
--- FAIL: TestTests (56.22s)
FAIL
make: *** [Makefile:69: functest] Error 1

```

alicefr replied:
Just to confirmed that I added an ugly sleep in the code:
```diff 
git diff
diff --git a/tests/vmi_configuration_test.go b/tests/vmi_configuration_test.go
index d3735596a..4263b8c59 100644
--- a/tests/vmi_configuration_test.go
+++ b/tests/vmi_configuration_test.go
@@ -999,7 +999,7 @@ var _ = Describe("Configurations", func() {
                                        guestMemReq := resource.MustParse(guestMemory)
                                        hugepagesVmi.Spec.Domain.Memory.Guest = &guestMemReq
                                }
-
+                               time.Sleep(30 * time.Second)
                                By("Starting a VM")
                                _, err = virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Create(hugepagesVmi)
                                Expect(err).ToNot(HaveOccurred())
```
and the tests are now passing.

I need some feedback from you, what is worthy to implement. The proper way should be to loop until the free_pages==nr_pages  and rserv_page = 0 before starting the next testcase. It requiers of course some additional functions

vladikr commented on tests/vmi_configuration_test.go:
```diff
@@ -945,8 +945,19 @@ var _ = Describe("Configurations", func() {
+
 				// Verify that the VM memory equals to a number of consumed hugepages
-				vmHugepagesConsumption := int64(totalHugepages-freeHugepages) * hugepagesSize.Value()
+				vmHugepagesConsumption := int64(totalHugepages-freeHugepages+resvHugepages) * hugepagesSize.Value()
```
@alicefr I think we only need to care about the reserved pages. I guess we don't need this calculation.
Let's just do `vmHugepagesConsumption := int64(resvHugepages) * hugepagesSize.Value()` and it all should just work... I hope :)

alicefr replied:
@vladikr I'm not sure. The example above https://github.com/kubevirt/kubevirt/pull/4027#issuecomment-680730415 and https://github.com/kubevirt/kubevirt/pull/4027#issuecomment-680735124 have a "wierd" page number because of the mismatch between MiB and MB. So, I was not actually using 32 MiB and 64 MiB. Let me redo the 2 examples with the right memory :):

Example with 32 Mi:
```bash
cat /proc/meminfo | grep Huge
AnonHugePages:    851968 kB
ShmemHugePages:        0 kB
HugePages_Total:      64
HugePages_Free:       54
HugePages_Rsvd:        6
HugePages_Surp:        0
Hugepagesize:       2048 kB
Hugetlb:          131072 kB
```

Example with 64 Mi:
```bash
cat /proc/meminfo | grep Huge
AnonHugePages:    849920 kB
ShmemHugePages:        0 kB
HugePages_Total:      64
HugePages_Free:       32
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:       2048 kB
Hugetlb:          131072 kB
```
Plus some extra ;) 

Example with 16Mi:
```bash
cat /proc/meminfo | grep Huge
AnonHugePages:    845824 kB
ShmemHugePages:        0 kB
HugePages_Total:      64
HugePages_Free:       56
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:       2048 kB
Hugetlb:          131072 kB
```
Example with 24 Mi
```bash
cat /proc/meminfo | grep Huge
AnonHugePages:    827392 kB
ShmemHugePages:        0 kB
HugePages_Total:      64
HugePages_Free:       54
HugePages_Rsvd:        2
HugePages_Surp:        0
Hugepagesize:       2048 kB
Hugetlb:          131072 kB

```

vladikr replied:
Interesting :)

Actually I didn't comment based on your example.

I've printed the values from the within the test on my machine:
```
freeHugepages:  62
resvHugepages:  32
totalHugepages-freeHugepages+resvHugepages:  34
pageSize:  2097152
vmHugepagesConsumption:  71303168
vmMemory:  67108864

------------------------------
• Failure [14.329 seconds]
Configurations
/root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:56
  VirtualMachineInstance definition
  /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:172
    [rfe_id:140][crit:medium][vendor:cnv-qe@redhat.com][level:component]with hugepages
    /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:912
      should consume hugepages 
      /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table.go:92
        [test_id:1671]hugepages-2Mi [It]
        /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table_entry.go:43

        Expected
            <int64>: 71303168
        to equal
            <int64>: 67108864

```

vladikr replied:
My assumption was that when the VMI starts, memfd will reserve the correct amount of pages for it and we test this at this point.
When I ran all the tests together with `vmHugepagesConsumption := int64(resvHugepages) * hugepagesSize.Value()` it passes, but I'm not sure if my assumption is true.

alicefr replied:
mmh :thinking: , might be that when I tried the examples above the VM was running since a while and then the reserved pages have been allocated and when you run the tests it was still in early stage. Not sure what is the behavior.

alicefr replied:
@vladikr did you set the memory with this change https://github.com/kubevirt/kubevirt/pull/4027#discussion_r477299771? Just to be sure

vladikr replied:
yeah, I just pulled your latest.

alicefr replied:
For me, that is what happens:
1-  Checking only the reserved pages when the VM is starting works because none of the pages has been allocated
2- The test that is failing is the second one because not all the hugepages from the previous test have been freed yet.
    This is why you got 34 pages and it worked only considering the reserved pages. However, as we can see from my 
    examples in the case of 64Mi all the pages have been allocated, but it is not the case for example for 32Mi.

alicefr replied:
It will be interesting to try if we change the memory to 32Mi of the first case, and check if the second one still passes only considering the reserved pages.

alicefr replied:
I tried to change it using only reserved pages but for me, it fails.

vladikr replied:
oh I see... it may be only a coincidence that it worked with 64Mi
Ok, then let's follow your suggestion to wait until `free_pages==nr_pages and rserv_page = 0`
I only hope that we will get to this condition in a reasonable time so we will keep it deterministic.

Perhaps you can just wrap it in an `Eventually` statement, like in 
https://github.com/kubevirt/kubevirt/blob/master/tests/vmi_configuration_test.go#L1030

alicefr replied:
@vladikr I'll figure out where to include the check :)

alicefr replied:
@vladikr do you have any suggestions on how the VM creation [here](https://github.com/kubevirt/kubevirt/blob/master/tests/vmi_configuration_test.go#L995) can be splitted in VM defined and then set the VM in running state later. I'd like to create the vmi pod, check if the hugepages have been freed, and then start the VM.

alicefr replied:
alternatively, we can just try multiple times to verify the [memory consumption](https://github.com/kubevirt/kubevirt/blob/master/tests/vmi_configuration_test.go#L1000). That is much simpler to implement. Right now, I'm having some issues with Eventually, because it doesn't retry a second time if the condition is not met. I need to verify why

vladikr replied:
@alicefr I think verifying consumption is an easier approach. Let me try your PR locally and see what's wrong with the `Eventually`

alicefr replied:
@vladikr I added a couple of prints locally, and the problem is that if the function `verifyHugepagesConsumption` doesn't return immediately true, then eventually doesn't retry a second time and it just times out. At least, it's what I've locally

vladikr replied:
ah, yes, I assume it might take some time for `tests.ExecuteCommandOnPod` to execute. Perhaps, just increase the timeout (maybe even to a minute), it should help.

alicefr replied:
Right, I've already tried to increase up to 2 min ;), but again the problem it's that it doesn't retry a second time. So, after 2 mins it times out

alicefr replied:
I also tried to measure how much time the function takes, it's more or less 300ms.

alicefr replied:
@vladikr without eventually, with a very basic loop it's working.
```diff
+                               for try := 0; !verifyHugepagesConsumption(); try++ {
+                                       if try == 5 {
+                                               break
+                                       }
+                                       time.Sleep(time.Second)
+                                       fmt.Printf("Try: %d \n", try)
+                               }
```

alicefr replied:
@vladikr moving the entire function inside the `Eventually` brackets solved the issue. I hope that now it's fine.

vladikr commented:
/ok-to-test

alicefr commented:
I need to check libvirt version in order to use `memfd`, some tests are failing because it's not supported

alicefr commented:
Add 61b16fd1017e53b742f97ae24d0ffb28697d50ee and 5b27b9e924879c65d33b6f9a2ab3976653838ec` to check libvirt version and set memfd accordingly. Let's see if the CI is happy now. If you don't like it, I can always remove them

vladikr commented:
Thanks @alicefr 
I think that the real problem is the 1.16 provider which is based on CentOS7, we should drop our support for it as soon as https://github.com/kubevirt/kubevirtci/pull/434 merges
After this should shouldn't have an issue with having a correct libvirt version for memfd.

I think what we need to do is to simply skip the hugepages test on 1.16 lane - we can move the libvirt version check to the functional test and skip the test until 1.19 is introduced - what do you think?

alicefr commented:
@vladikr yes, this can be a workaround. Do you have only have centos:7 for testing? If this is the case, I can try to skip the test for this version

vladikr commented:
> @vladikr yes, this can be a workaround. Do you have only have centos:7 for testing? If this is the case, I can try to skip the test for this version

Normally, we use the same virt-launcher image across all the providers, but it seems that 1.16 provider is much different. it is based on centos7 and looks like it uses an older virt-launcher image.
I think it was a mistake to go this route but now we just need to wait until it is deprecated.

alicefr commented:
@vladikr do you think I can simply skip the hugepages tests if k8s cluster is 1.16? Getting libvirt version from the tests is tedious because I need to get the version from the running to compute container

alicefr commented:
I'm skipping now the hugepage tests based on the k8s version. Let's see if the CI is happy now :)

alicefr commented:
Right now, the hugepages tests are correctly skipped for version 1.16 (see [here](https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/4027/pull-kubevirt-e2e-k8s-1.16/1301769978910871552#1:build-log.txt%3A5266)), but still some tests are failing. Not sure if it is due to this pr (see [here](https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/4027/pull-kubevirt-e2e-k8s-1.17/1301542735043366912#1:build-log.txt%3A3763))

vladikr commented:
/retest

alicefr commented:
@vladikr could you please retrigger the last test? Is there a way to be part of kubevirt community and self trigger the tests:)?

vladikr commented:
/test all

vladikr commented:
@alicefr , one concern that @fabiand and @davidvossel raised is that we will need a way to fallback to consuming hugepages without memfd. This is needed since we are not sure if there is any kind of overhead associated with memfd and we need an escape hatch until we are ceraitn that there isn't.
Without changing the VMI API and the virt-luncher <-> virt-handler api, I would suggest simply to use an annotaion, if it's set do not set memfd and numa elements. I think it should be straight forward, something like [1]

What do you think?

[1] https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-launcher/virtwrap/api/converter.go#L1291

alicefr commented:
@vladikr yes, it makes sense if there are any concerns. We can always remove the check once it becomes more stable. Do you think that memfd can be the default, and we can turn it off or the other way round? I'll check in these days how to add the annotation.

alicefr commented:
@vladikr I added the annotation. Please, check the last 2 commits. If you use a similar example, memfd is not used:
```yaml
---
apiVersion: kubevirt.io/v1alpha3
kind: VirtualMachineInstance
metadata:
  annotations:
    kubevirt.io/memfd: "false"
  labels:
    special: vmi-hugepages
  name: vmi-hugepages
spec:
  domain:
    memory:
      hugepages:
        pageSize: 2Mi
    devices:
      disks:
      - disk:
          bus: virtio
        name: containerdisk
    machine:
      type: ""
    resources:
      requests:
        memory: 32Mi
  terminationGracePeriodSeconds: 0
  volumes:
  - containerDisk:
      image: registry:5000/kubevirt/cirros-container-disk-demo:devel
    name: containerdisk           
```
I'm not sure, what is the correct way to update the `staging/src/kubevirt.io/client-go`. I guess, I need to open a PR in https://github.com/kubevirt/client-go. Right now, I simply modified it directly in the code. So, before merging it, I need to fix that. Additionally, I changed the behavior for k8s 1.16 tests. Now, I'm setting the annotation `kubevirt.io/memfd=false` and the tests should fall back to the previous behavior.

vladikr commented:
> I'm not sure, what is the correct way to update the `staging/src/kubevirt.io/client-go`. I guess, I need to open a PR in https://github.com/kubevirt/client-go. Right now, I simply modified it directly in the code. So, before merging it, I need to fix that. Additionally, I changed the behavior for k8s 1.16 tests. Now, I'm setting the annotation `kubevirt.io/memfd=false` and the tests should fall back to the previous behavior.

@alicefr there is no need. just add a new type somewhere here: https://github.com/kubevirt/kubevirt/blob/master/staging/src/kubevirt.io/client-go/api/v1/types.go#L511
and run `make && make generate` then you can use this new type `v1.YourType` in the converter

vladikr commented:
> @vladikr yes, it makes sense if there are any concerns. We can always remove the check once it becomes more stable. Do you think that memfd can be the default, and we can turn it off or the other way round? I'll check in these days how to add the annotation.

Thanks @alicefr . Yes, I think memfd can be default, we will just fall back to the previous behavior if the annotation is set.

alicefr commented:
> > I'm not sure, what is the correct way to update the `staging/src/kubevirt.io/client-go`. I guess, I need to open a PR in https://github.com/kubevirt/client-go. Right now, I simply modified it directly in the code. So, before merging it, I need to fix that. Additionally, I changed the behavior for k8s 1.16 tests. Now, I'm setting the annotation `kubevirt.io/memfd=false` and the tests should fall back to the previous behavior.
> 
> @alicefr there is no need. just add a new type somewhere here: https://github.com/kubevirt/kubevirt/blob/master/staging/src/kubevirt.io/client-go/api/v1/types.go#L511
> and run `make && make generate` then you can use this new type `v1.YourType` in the converter

@vladikr then, it's what I did

vladikr reviewed (approved):
Great! Thanks @alicefr 
I'll give a chance to others to review this as well before merging it.
/lgtm

alicefr commented:
I squashed the 2 test commits and removed the test skip to have a cleaner structure

jean-edouard commented:
Is there a downside to just switching to memfd, without even an option to change it?
That way we wouldn't have to support multiple backends, and more importantly, we could remove the `hugetlbfs_t` rules from our custom SELinux policy.

alicefr commented:
@jean-edouard it's not clear yet. Please, have a look to Vladik's [comment](https://github.com/kubevirt/kubevirt/pull/4027#issuecomment-688585396). Additionally, the provider for k8s 1.16 has a tool old libvirt version that doesn't support memfd. This might be also a way to fallback to the previous behavior if memfd is not supported.

vladikr commented:
> Is there a downside to just switching to memfd, without even an option to change it?
> That way we wouldn't have to support multiple backends, and more importantly, we could remove the `hugetlbfs_t` rules from our custom SELinux policy.

I would also like to remove `hugetlbfs_t` from our SELinux policy, but I don't this we should do it at this stage. We can remove it when we are comfortable enough with just memfd.

jean-edouard commented:
Sounds good, sorry I missed that comment when catching up with the history of this PR.
I guess we can switch to memfd by default after we get rid of the k8s-1.16 test lane.

vladikr commented:
/approve

jean-edouard commented:
@alicefr I think this is worth a release note in the PR description.

alicefr commented:
@jean-edouard done
